modules/renderer/features.js

```diff
@@ -94,6 +96,44 @@ export function rendererFeatures(context) {
         };
     }
 
+    defineRule('date_range', function isWithinRange(tags) {
+      // dates a are a mix of yyyy-mm-dd yyyy-mm and yyyy, so pad them out
+      // then use DecimalDate JS https://github.com/OpenHistoricalMap/decimaldate-javascript/ to properly handle negative BCE dates
+      var dateToNumber = function(date) {
+        if (! date) return NaN;
+
+        let paddeddate;
+        if (date.match(/^\-?\d\d\d\d\-\d\d\-\d\d$/)) {
+          paddeddate = date; // perfect yyyy-mm-dd
+        } else if (date.match(/^\-?\d\d\d\d\-\d\d$/)) {
+          const ym = date.match(/^\-?(\d\d\d\d)\-(\d\d)$/);
+          const d = (new Date(parseInt(ym[1]), parseInt(ym[2]), 0)).getDate();
+          paddeddate = date + `-${String(d).padStart(2, '0')}`; // yyyy-mm, add -dd
+        } else if (date.match(/^\-?\d\d\d\d$/)) {
+          paddeddate = date + '-12-31'; // yyyy, add -mm-dd
+        } else if (date.match(/^\-?\d{1,3}$/)) {
+          paddeddate = date.padStart(4, '0') + '-12-31';
+        }
+
+        return decimaldate.iso2dec(paddeddate);
+      };
+
+      // entity's start & end date
+      // filtering date range from the on-screen controls
+      const entityRange = {
+        'start_date': dateToNumber(tags.start_date),
+        'end_date': dateToNumber(tags.end_date)
+      };
+      const selectedRange = {
+        'start_date': dateToNumber(context.features().dateRange[0]),
+        'end_date': dateToNumber(context.features().dateRange[1])
+      };
+
+      // out of range = feature started after range ends, or feature ends before range starts
+      const withinrange = !(selectedRange.start_date > entityRange.end_date) && !(selectedRange.end_date < entityRange.start_date);
```

Comment:
#137 added a `utilDatesOverlap()` method to util.js that would allow this to say something like:

```js
const withinrange = utilDatesOverlap(selectedRange, entityRange);
```

`utilDatesOverlap()` assumes dates are well-formed so that it can simply perform some string comparisons. I assume these string comparisons would be more performant than converting to decimal dates first. If a date is malformed, it treats it like the empty string and assumes the feature stretches to the beginning or end of time.

Reply:
I have changed to use `utilNormalizeDateString()` and `utilDatesOverlap()`